Convert misc chapters to prodn, update syntax (#12936)
coq/coq: doc/sphinx/addendum/program.rst

   typechecking.

.. attr:: program

>    :name: program; Program

Author: It seems harmless to include and it is good for completeness.  So I would keep it.

>    :name: program; Program

Reviewer: Sure.

Program Definition
~~~~~~~~~~~~~~~~~~

>    A :cmd:`Definition` command with the :attr:`program` attribute types

Reviewer: What about removing the indentation?

   the value term in Russell and generates proof
   obligations. Once solved using the commands shown below, it binds the
   final |Coq| term to the name :n:`@ident` in the environment.

   .. exn:: @ident already exists.
      :name: @ident already exists. (Program Definition)

>       :undocumented:

Reviewer: Let's remove that. This is redundant with the exact same error message that is already shown for the `Definition` command.

   :n:`Program Definition @ident : @type := @term`

   Interprets the type :n:`@type`, potentially generating proof
   obligations to be resolved. Once done with them, we have a |Coq|
   type :n:`@type__0`. It then elaborates the preterm :n:`@term` into a |Coq|
   term :n:`@term__0`, checking that the type of :n:`@term__0` is coercible to
   :n:`@type__0`, and registers :n:`@ident` as being of type :n:`@type__0` once the
   set of obligations generated during the interpretation of :n:`@term__0`
   and the aforementioned coercion derivation are solved.

    .. exn:: In environment … the term: @term does not have type @type. Actually, it has type ...

>        :undocumented:

Reviewer: This error message does not actually exists. There is no Program-specific error message in this case, so let's just remove this as well.

   :n:`Program Definition @ident {* @binder } : @type := @term`

   Is equivalent to:

>    :n:`Program Definition @ident : forall {* @binder }, @type := fun {* @binder } => @term`.

Reviewer: This is not specific to `Program Definition`, so I don't think it's useful to repeat it here. Let's also remove it.

Program Definition
~~~~~~~~~~~~~~~~~~

.. todo PR need to understand this better.  Is there a reason this only has

>       a subset of the Definition syntax?  Maybe need to ask Matthieu

Author: Check what I did.

Program Fixpoint
~~~~~~~~~~~~~~~~

.. cmd:: Program Fixpoint @fix_definition {* with @fix_definition }

   The optional :n:`@fixannot` annotation can be one of:

   + :g:`measure f R` where :g:`f` is a value of type :g:`X` computed on
     any subset of the arguments and the optional term
     :g:`R` is a relation on :g:`X`. :g:`X` defaults to :g:`nat` and :g:`R`
     to :g:`lt`.

>    + :g:`wf R x` which is equivalent to :g:`measure x R`.

Reviewer: Let's keep the description of the various options for `fixannot` here (except you can move it below the first example if you want).

>    A :cmd:`Fixpoint` command with the :attr:`program` attribute may also generate obligations. It works

Reviewer: Same suggestion about removing the indentation.

Program Fixpoint
~~~~~~~~~~~~~~~~

.. cmd:: Program Fixpoint @fix_definition {* with @fix_definition }

>    .. todo PR: move measure and wf info to inductive.html#recursive-functions-fix?

Reviewer 2: The `wf` annotation is treated the same but for measures Program differs indeed.

>    .. todo PR: move measure and wf info to inductive.html#recursive-functions-fix?

Author: @mattam82 Can you elaborate on what the difference is here?  Just a few lines below, I have text that says ":g:`wf R x` which is equivalent to :g:`measure x R`." which seems inconsistent with you statement (that wf is the same but measure is difterent).

>    .. todo PR: move measure and wf info to inductive.html#recursive-functions-fix?

Reviewer: What @mattam82 said is that `wf` annotation is treated the same way in `Function` and `Program` and that `measure` means something different in `Function` and `Program`. In `Program`, `wf` means the same up to the order of the arguments. This is not inconsistent with the fact that `measure` means something else in `Function`.

BTW @mattam82, this is kind of confusing, so don't you think that `measure` should be disallowed with `Program`?

>    .. todo PR: move measure and wf info to inductive.html#recursive-functions-fix?

Reviewer 2: `measure` is nicer in Program as it allows the user not to pack the arguments in a tuple but can mention all of them.
One can write `measure (length l + length l')` not having to mention that the relation should be `Nat.lt`.

>    .. todo PR: move measure and wf info to inductive.html#recursive-functions-fix?

Author: I need a bit more explanation of `measure`, `wf` and `struct` (not described in the doc) before I can say how they're different with or without `Program`.  I don't have a clear grasp on their meaning.  I take it `measure` has to decrease for each recursion in the definition (maybe not the best words), but I have no clue how that differs from the other two.  `struct` means the number of # symbols in specified variable must decrease for each recursion?  No idea how tuples fit into this.  A couple very simple examples would help.

> `wf` means the same up to the order of the arguments

Meaning ignoring the order of the arguments?  Not so familiar with the "up to" idiom.

>    .. todo PR: move measure and wf info to inductive.html#recursive-functions-fix?

Reviewer: `struct` is for structural recursion. It means you can only call the recursive function on strict subterms (obtained by pattern-matching) of your initial argument. This is the simplest form of recursion, the only one accepted by the kernel.

>Meaning ignoring the order of the arguments?

In this case, I mean that the order of arguments is inverted between the two syntax.

Program Lemma
~~~~~~~~~~~~~

>    A :cmd:`Lemma` command with the :attr:`program` attribute uses the Russell

Reviewer: Same suggestion about removing the indentation.

   language to type statements of logical
   properties. It generates obligations, tries to solve them
   automatically and fails if some unsolved obligations remain. In this
   case, one can first define the lemma’s statement using :cmd:`Definition`
   and use it as the goal afterwards. Otherwise the proof
   will be started with the elaborated version as a goal. The

>    :g:`Program` attribute can similarly be used as a prefix for

Reviewer: ```suggestion
   :attr:`Program` attribute can similarly be used with
```

   automatically, whether to solve them or when starting to prove one,
   e.g. using :cmd:`Next Obligation`.

   This command supports the :attr:`local` and :attr:`global` attributes.
   :attr:`local` makes the setting last only for the current
   module. :attr:`local` is the default inside sections while :attr:`global`
   otherwise.

.. cmd:: Show Obligation Tactic

   Displays the current default tactic.

.. cmd:: Obligations {? of @ident }

   Displays all remaining obligations.

.. cmd:: Obligation @natural {? of @ident } {? : @type {? with @ltac_expr } }

   Start the proof of obligation :token:`natural`.

.. cmd:: Next Obligation {? of @ident } {? with @ltac_expr }

   Start the proof of the next unsolved obligation.

.. cmd:: Solve Obligations {? {? of @ident } with @ltac_expr }

   Tries to solve each obligation of :token:`ident` using the given :token:`ltac_expr` or the default one.

.. cmd:: Solve All Obligations {? with @ltac_expr }

   Tries to solve each obligation of every program using the given
   tactic or the default one (useful for mutually recursive definitions).

.. cmd:: Admit Obligations {? of @ident }

>    Admits all obligations (of ``ident``).

Reviewer: ```suggestion
   Admits all obligations (of :token:`ident`).
```

obligations (e.g. when defining mutually recursive blocks). The
optional tactic is replaced by the default one if not specified.

.. cmd:: Obligation Tactic := @ltac_expr
   :name: Obligation Tactic

   Sets the default obligation solving tactic applied to all obligations
   automatically, whether to solve them or when starting to prove one,
   e.g. using :g:`Next`.

   This command supports the :attr:`local` and :attr:`global` attributes.
   :attr:`local` makes the setting last only for the current
   module. :attr:`local` is the default inside sections while :attr:`global`
   otherwise.

.. cmd:: Show Obligation Tactic

   Displays the current default tactic.

.. cmd:: Obligations {? of @ident }

   Displays all remaining obligations.

> .. cmd:: Obligation @int {? of @ident } {? : @term {? with @ltac_expr } }

Reviewer 3: This is `@natural` since #12979, a rebase should fix that.

   Start the proof of obligation :token:`int`.

.. cmd:: Next Obligation {? of @ident } {? with @ltac_expr }

   Start the proof of the next unsolved obligation.

> .. cmd:: Solve Obligations {? {? of @ident } with @ltac_expr }

Author: That matches the grammar; it doesn't allow `of @ident` without `with @ltac_expr`.

```
| "Solve" "Obligations" "of" ident "with" tactic
| "Solve" "Obligations" "with" tactic
| "Solve" "Obligations"
```

though that doesn't get at "why", which I have no opinion on.

> .. cmd:: Solve Obligations {? {? of @ident } with @ltac_expr }

Reviewer: Yes, I did not think that the tool was mistaken. It's thus more a question for @mattam82.

> .. cmd:: Solve Obligations {? {? of @ident } with @ltac_expr }

Author: This looks like an oversight.  I'll submit a PR to add it.
